src/java.base/share/classes/jdk/internal/classfile/impl/AnnotationReader.java

```diff
@@ -283,20 +281,78 @@ private static int skipTypeAnnotation(ClassReader classReader, int p) {
     }
 
     public static void writeAnnotation(BufWriterImpl buf, Annotation annotation) {
-        // handles annotations and type annotations
         // TODO annotation cleanup later
         ((Util.Writable) annotation).writeTo(buf);
     }
 
-    public static void writeAnnotations(BufWriter buf, List<? extends Annotation> list) {
-        // handles annotations and type annotations
+    public static void writeAnnotations(BufWriter buf, List<Annotation> list) {
         var internalBuf = (BufWriterImpl) buf;
         internalBuf.writeU2(list.size());
         for (var e : list) {
             writeAnnotation(internalBuf, e);
         }
     }
 
+    private static int labelToBci(LabelContext lr, Label label, TypeAnnotation ta) {
+        //helper method to avoid NPE
+        if (lr == null) throw new IllegalArgumentException("Illegal targetType '%s' in TypeAnnotation outside of Code attribute".formatted(ta.targetInfo().targetType()));
+        return lr.labelToBci(label);
+    }
+
+    public static void writeTypeAnnotation(BufWriterImpl buf, TypeAnnotation ta) {
```

Comment:
Is there any reason to move writeTypeAnnotation from UnboundAttribute?

Reply:
This is for consistency with reading annotations: they are now defined in the same file.